lib/linked_list.rb

```diff
@@ -18,71 +21,213 @@ def initialize
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
+      if @head.nil?
+        @head = Node.new(value)
+      else 
+        current = @head
+        first = Node.new(value)
+        first.next = current
+        current.prev = first
+        @head = first
+      end 
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      current = @head
+      if @head.nil?
+        return false
+      else 
+        while current.next != nil
+          if current.data == value
+            return true
+          else 
+            current = current.next
+          end
+        end
+      end
+      if current.data == value
+        return true
+      else 
+        return false
+      end 
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def find_max
-      raise NotImplementedError
+      max = -1000000 
+      current = @head
+      if @head.nil?
+        return nil
+      elsif @head.next.nil?
+        return @head.data
+      else 
+        while current.next != nil
+          if current.data > max
+            max = current.data
+          end
+          current = current.next
+        end
+      end
+
+      if current.data > max
+        max = current.data
+      end
+
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      min = 1000000
+      current = @head 
+      if @head.nil?
+        return nil
+      elsif @head.next.nil?
+        return @head.data
+      else 
+        while current.next != nil
+          if current.data < min
+            min = current.data
+          end
+          current = current.next
+        end 
+      end 
+
+      if current.data < min
+        min = current.data
+      end
+
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      length = 1
+      current = @head
+      if current.nil?
+        return 0
+      elsif current.next.nil?
+        return 1
+      else 
+        while current.next != nil
+          length +=1
+          current = current.next
+        end
+      end
+      return length

```

Comment:
This could be simplified a bit:

```suggestion
     len = 0
     current = @head
     until current.nil?
       len += 1
       current = current.next
     end
     return len
       
```